platform/windows/context_gl_windows.cpp

```diff
@@ -63,16 +66,76 @@ int ContextGL_Windows::get_window_height() {
 	return OS::get_singleton()->get_video_mode().height;
 }
 
+bool ContextGL_Windows::should_vsync_via_compositor() {
+
+	if (OS::get_singleton()->is_window_fullscreen()) {
+		return false;
+	} else {
+		// Note: All Windows versions supported by Godot (Vista and later)
+		// have a compositor.  It can be disabled on earlier Windows versions.
+		BOOL dwm_enabled;
+
+		if (SUCCEEDED(DwmIsCompositionEnabled(&dwm_enabled))) {
+			return dwm_enabled;
+		}
+
+		return false;
+	}
+}
+
 void ContextGL_Windows::swap_buffers() {
 
+	if (use_vsync) {
+		bool sync_via_compositor = should_vsync_via_compositor();
+
+		if (sync_via_compositor) {
+			DwmFlush();
+		}
+
+		if ((sync_via_compositor && swap_interval > 0) ||
+				(!sync_via_compositor && swap_interval == 0)) {
+
+			// One of the following just occurred...
+			//
+			// 1) Vertical syncing via the compositor has just been enabled.
+			//    This will happen in the following cases:
+			//      a) When switching from full screen mode to windowed mode
+			//         and the OS compositor is enabled.
+			//      b) When, in windowed mode, the user enables the compositor--
+			//         by switching themes, for example.
+			//
+			//    In this case turn vsync via the compositor on and disable it
+			//    for OpenGL.
+			//
+			// 2) Vertical syncing via the compositor has just been disabled.
+			//    This will happen in the following cases:
+			//      a) When switching from windowed mode to full screen mode
+			//         regardless whether the OS compositor is enabled or not.
+			//      b) When, in windowed mode, the user disables the
+			//         compositor--by switching themes, for example.
+			//
+			//    In this case turn vsync via OpenGL on and disable it for the
+			//    compositor.
+
+			set_use_vsync(true);
+		}
+	}
+
 	SwapBuffers(hDC);
 }
 
 void ContextGL_Windows::set_use_vsync(bool p_use) {
 
+	if (p_use && should_vsync_via_compositor()) {
+		swap_interval = 0; // vsync will be done via DwmFlush().
```

Comment:
It seems hacky to rely on the swap interval to define whether vsync should be handled by the compositor. I'd use a boolean like `vsync_via_compositor` to make things explicit.

Reply:
I reworked it as you suggested.